[VC] Revise README.md (#1313)
kubernetes-sigs/multi-tenancy: incubator/virtualcluster/README.md

> With VirtualCluster, each tenant is assigned a dedicated tenant master, which is a upstream Kubernetes distribution.

Reviewer: ```suggestion
With VirtualCluster, each tenant is assigned a dedicated tenant control plane, which is a upstream Kubernetes distribution.
```

Please follow the [instructions](./doc/demo.md) to install VirtualCluster in your local K8s cluster.

## Abstraction

> In VirtualCluster, tenant master owns the source of the truth for the specs of all the synced objects. 

Reviewer: ```suggestion
In VirtualCluster, tenant control plane owns the source of the truth for the specs of all the synced objects. 
```

> The exceptions are persistence volume and storage class objects whose source of the truth is the super cluster.

Reviewer: ```suggestion
The exceptions are persistence volume, storage class, and priority class resources whose source of the truth is the super cluster.
```

> The syncer along with the super cluster acts like a resource provider controller which updates the synced

Reviewer: What do you mean by "acts like a resource provider controller"?

> object's status in each tenant master. This abstraction model means the following assumptions:

Reviewer: ```suggestion
object's status in each tenant control plane. This abstraction model means the following assumptions:
```

> - The synced object spec should not be altered by any arbitrary controller in the super cluster.

Reviewer: ```suggestion
- The synced object spec _SHOULD_ not be altered by any arbitrary controller in the super cluster.
```

> - Tenant master owns the life cycle management for the synced object. The synced objects should not be

Reviewer: ```suggestion
- Tenant control plane owns the lifecycle management for the synced object. The synced objects should not be
```

>   managed by any controllers (say StatefulSet for example) in the super cluster.

Reviewer: ```suggestion
  managed by any controllers (e.g. `StatefulSet` for example) in the super cluster.
```

If any of the above assumptions is violated, VirtualCluster may not work as expected. Note that this 
does not mean that a cluster administrator cannot install webhooks, for example, a sidecar webhook, 

> in the super cluster. Those webhooks still work but the cluster administrator admits that the changes are going

Reviewer: ```suggestion
in the super cluster. Those webhooks will still work but the changes are going
```

> to be hidden to tenants. Alternatively, those webhooks can be installed in tenant masters so that

Reviewer: ```suggestion
to be hidden to tenants. Alternatively, those webhooks can be installed in tenant control planes so that
```

tenants will be aware of all changes.

## Limitations

Ideally, tenants should not be aware of the existence of the super cluster in most cases. 

> There are still some noticeable differences comparing a tenant master and a normal kubernetes cluster.

Reviewer: ```suggestion
There are still some noticeable differences comparing a tenant control plane and a normal kubernetes cluster.
```

> - In tenant master, node objects only show up after tenant Pods are created. The super cluster

Reviewer: ```suggestion
- In the tenant control plane, node objects only show up after tenant Pods are created. The super cluster
```

>   node topology is not fully exposed in the tenant master. This means VirtualCluster does not support

Reviewer: ```suggestion
  node topology is not fully exposed in the tenant master. This means the VirtualCluster does not support
```

>   `DaemonSet` alike workloads in tenant master. Currently, the syncer controller rejects a newly

Reviewer: ```suggestion
  `DaemonSet` alike workloads in tenant control plane. Currently, the syncer controller rejects a newly
```

  created tenant Pod if its `nodename` has been set in the spec. 

> - The syncer controller manages the life cycle of the node objects in tenant master but

Reviewer: ```suggestion
- The syncer controller manages the lifecycle of the node objects in tenant control plane but
```

>   it does not update the node lease objects in order to reduce network traffics. As a result,

Reviewer: ```suggestion
  it does not update the node lease objects in order to reduce network traffic. As a result,
```

>   it is recommended to increase the tenant master node controller `--node-monitor-grace-period` 

Reviewer: ```suggestion
  it is recommended to increase the tenant control plane node controller `--node-monitor-grace-period` 
```

  parameter to a larger value ( >60 seconds, done in the sample clusterversion
  [yaml](config/sampleswithspec/clusterversion_v1_nodeport.yaml) already).

> - Coredns is not tenant-aware. Hence, tenants should install coredns in tenant masters if DNS is required.

Reviewer: ```suggestion
- Coredns is not tenant-aware. Hence, tenants should install coredns in the tenant cluster if DNS is required.
```

The DNS service should be created in the `kube-system` namespace using the name `kube-dns`. The syncer controller can then

> recognize the DNS service cluster IP in super cluster and inject it into Pod spec `dnsConfig`.

Reviewer: ```suggestion
recognize the DNS service's `ClusterIP` in super cluster and inject it into any Pod `spec.dnsConfig`.
```

- The cluster IP field in the tenant Service Spec is a bogus value. If any tenant controller requires the
actual cluster IP that takes effect in the super cluster nodes, a special handling is required. 
The workaround usually is going to be a simple code change in the controller. This 

> [document](./doc/tenant-dns.md) shows an example for coredns.

Reviewer: I wonder if in this section we can write out the annotation so that someone could look up the clusterIP without having to read the tenant-dns doc?

- VirtualCluster does not support tenant PersistentVolumes. All PVs and Storageclasses are provided by the super cluster.

VirtualCluster passes most of the Kubernetes conformance tests. One failing test asks for supporting
`subdomain` which cannot be easily done in the VirtualCluster.

## FAQ

### Q: What is the difference between VirtualCluster and multi-cluster solution?

One of the primary design goals of VirtualCluster is to improve the overall resource utilization

> in super cluster by allowing multiple tenants to share the node resource in an isolated manner. 

Reviewer: ```suggestion
of a super cluster by allowing multiple tenants to share the node resources in a control plane isolated manner. 
```

> A multi-cluster solution can achieve the same isolation goal but the overall resource utilization
> is going to be suboptimal.

Reviewer: ```suggestion
A multi-cluster solution can achieve the same isolation goal but resources won't be shared causing nodes to have lower utilization.
```

> ### Q: Can the tenant master run its own scheduler?

Reviewer: ```suggestion
### Q: Can the tenant controller run its own scheduler?
```

VirtualCluster was primarily designed for serverless use cases where users normally do not have
scheduling preferences. Using the super cluster scheduler can much easily
achieve good overall resource utilization. For these reasons, 
VirtualCluster does not support tenant scheduler. It is technically possible
to support tenant scheduler by exposing some of the super cluster nodes directly in

> tenant master. Those nodes have to be dedicated to the tenant to avoid any scheduling

Reviewer: ```suggestion
tenant control plane. Those nodes have to be dedicated to the tenant to avoid any scheduling
```

conflicts. This type of tenant should be exceptional.

### Q: What is the difference between Syncer and Virtual Kubelet? 

They have similarities. In some sense, the syncer controller can be viewed as the replacement of a virtual

> kubelet in cases where the resource provider of the virtual kubelet is a Kubernetes cluster. The syncer certainly does more
> though. It maintains the one to one mapping between a virtual node in tenant master and a real node

Reviewer: ```suggestion
kubelet in cases where the resource provider of the virtual kubelet is a Kubernetes cluster. The syncer maintains the one to one mapping between a virtual node in tenant control plane and a real node
```

> in super cluster. It preserves the Kubernetes API compatibility as completely as possible. It provides fair queuing to
> mitigate tenant interference. It is a more native implementation for tenant abstraction.

Reviewer: ```suggestion
in the super cluster. It preserves the Kubernetes API compatibility as closely as possible. Additionally it provides fair queuing to
mitigate tenant interference. It is a more native implementation for tenant abstraction.
```

Reviewer: What do you think about instead of `interference` we used `contention`?

Also, what do you mean by "It is a more native implementation for tenant abstraction."?
